File: RCode/Chapter3/imputationMissingValues.R
##########################################################################
#
#   Mark Hoogendoorn & Burkhardt Funk (2017)
#   Machine Learning for the Quantified Self
#   Springer
#   ./chapter3/imputationMissingValues.R
# 
##########################################################################

# Imputation is covered in different R packages, one is mice - it offers lots of methods
# see methods(mice)

imputeMean = function(df,col) {
  df[is.na(df[,col]),col] = mean(df[,col],na.rm = TRUE)
  return(df)
}

imputeMedian = function(df,col) {
  df[is.na(df[,col]),col] = median(df[,col],na.rm = TRUE)
  return(df)
}

# for the following function there are other existing functions in the zoo package like na.spline. 
# Other packages such as imputeTS also supports imputation for time series
imputeInterpolation = function(df,col) {
  dat = zoo(df[,col])
  dat = ifelse(is.nan(dat),NA,dat)
  # fill initial and trailing NA
  if (is.na(dat[1])) dat[1] = dat[which.max(!is.na(dat))]
  if (is.na(dat[length(dat)])) dat[length(dat)] = dat[max((!is.na(dat))*(1:length(dat)))]
  # interpolate
  df[,col] = na.approx(dat)
  return(df)
}

File: Python3Code/Chapter3/ImputationMissingValues.py
##############################################################
#                                                            #
#    Mark Hoogendoorn and Burkhardt Funk (2017)              #
#    Machine Learning for the Quantified Self                #
#    Springer                                                #
#    Chapter 3                                               #
#                                                            #
##############################################################
import pandas as pd


class ImputationMissingValues:

    @staticmethod
    def impute_mean(dataset: pd.DataFrame, col: str) -> pd.DataFrame:
        """
        Impute the mean value in case if missing data.

        :param dataset: Dataset with missing values.
        :param col: Column containing missing values that will be filled.
        :return: Dataframe with imputed values in col.
        """

        dataset[col] = dataset[col].fillna(dataset[col].mean())
        return dataset

    @staticmethod
    def impute_median(dataset: pd.DataFrame, col: str) -> pd.DataFrame:
        """
        Impute the median value in case if missing data.

        :param dataset: Dataset with missing values.
        :param col: Column containing missing values that will be filled.
        :return: Dataframe with imputed values in col.
        """

        dataset[col] = dataset[col].fillna(dataset[col].median())
        return dataset

    @staticmethod
    def impute_interpolate(dataset: pd.DataFrame, col: str) -> pd.DataFrame:
        """
        Interpolate the dataset based on previous and next values in case of missing values.

        :param dataset: Dataset with missing values.
        :param col: Column containing missing values that will be filled.
        :return:
        """

        dataset[col] = dataset[col].interpolate()
        # And fill the initial data points if needed:
        dataset[col] = dataset[col].fillna(method='bfill')
        return dataset
